### ISSUE TITLE
Should System.Text.Json replace Newtonsoft.Json?

### ISSUE THREAD

**Issue by perason** [axuno/SmartFormat#126]
Given that ASP.NET Core 3 now has its own Json namespace (System.Text.Json), and with it provides high-performance, low-allocating process capabilities, do you have any plans to remove the Newtonsoft.Json dependencies any time soon?

Am sure that if you did, you would gain a performance increase for this great library (and MS themselves were able to gain 1.3x – 5x speed up).

https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-apis/

**Comment by axunonb** [comment]
Hi
Absolutely! This is something to be considered. At the being ```System.Text.Json``` [is not a drop-in replacement](https://medium.com/@samichkhachkhi/system-text-json-vs-newtonsoft-json-d01935068143) for ```Newtonsoft.Json```. So it's not only about performance, but also other consequences for existing projects using ```SmartFormat```. 
Are you suggesting to introduce a breaking change by migrating to ```System.Text.Json```, or to continue to support ```Newtonsoft.Json```, just like Microsoft are doing (for good reasons)?

**Comment by perason** [comment]
I am aware of it not being a dop-in-replacement, though I found it take care of most of my needs, but its `PopulateObject`, for which I got some help [asking this question at Stackoverflow](https://stackoverflow.com/questions/56835040/net-core-3-0-jsonserializer-populate-existing-object).

I haven't parsed SmartFormat's own code, just noticed its dependencies when installing it, so don't know how heavily it is used. If only for standard serializations, it should be easily replaced.

Whether introducing a _breaking change_ only you can decide, but if you try it out and find it boost your library, I now I would appreciate an alternative.

At first I think I would have added an option to the settings, where someone like me could opt-in for `System.Text.Json`.

**Comment by axunonb** [comment]
I see, thanks, also for SO link which is a good suggestion. JSON is not heavily integrated into ```SmartFormat```, it's just another data source. Would be interested to hear what other dev ops think.

**Comment by imtrobin** [comment]
There are parts which is mot yet ready to replace newtonjson

**Comment by perason** [comment]
> There are parts which is mot yet ready to replace newtonjson

Yes, there are, and very likely always will (and Microsoft have stated as well), still, I personally believe that for a library like this, performance could be a deal breaker (it is for me), where performance is more important than being able to do a huge lots of tricks....and I only ask for the ability to select which "Json" library it uses.

**Comment by axunonb** [comment]
With commit https://github.com/axuno/SmartFormat/commit/6724b06b3a06edea6ad005fff9f4506c16974884 we have added support ```System.Text.Json``` besides ```Newtonsoft.Json```. For ```Smart.Format``` it shouldn't matter, whether we consider ```System.Text.Json``` as feature-complete or not. The downside is, that support for .NetFramework 4.5x must be dropped.
You can now do something like that:
```csharp
using System.Text.Json;
[...]
var jElement = JsonDocument.Parse("some-json-string").RootElement;
var result = Smart.Format("{SomePlaceHolder}", jElement);
```
@perason, @imtrobin: Would like to have a look, whether this is what you had in mind?

**Comment by perason** [comment]
It might, so thanks for now, though have no time to dive more into it at the moment.

**Comment by axunonb** [comment]
Closed with release [v2.5.1](https://github.com/axuno/SmartFormat/releases/tag/v2.5.1.0)
